internal/controller/storageclient_controller.go

```diff
@@ -309,26 +416,125 @@ func (r *StorageClientReconciler) reconcilePhases() (ctrl.Result, error) {
 
-		utils.AddAnnotation(r.storageClient, storageClaimProcessedAnnotationKey, "true")
-		if err := r.update(r.storageClient); err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to update StorageClient with claim processed annotation: %v", err)
-		}
-	}
+				utils.AddLabels(storageClass, eResource.Labels)
+				utils.AddAnnotations(storageClass, eResource.Annotations)
+				return nil
+			}); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to create or update StorageClass: %s", err)
+			}
+		case "VolumeSnapshotClass":
+			snapshotClass := &snapapi.VolumeSnapshotClass{}
+			snapshotClass.Name = eResource.Name
+			if err := utils.CreateOrReplace(r.ctx, r.Client, snapshotClass, func() error {
+				if err := json.Unmarshal(eResource.Data, snapshotClass); err != nil {
+					return fmt.Errorf("failed to unmarshal snapshotclass configuration response: %v", err)
+				}
+				snapshotClass.Name = eResource.Name
+				removeStorageClaimAsOwner(snapshotClass)
+				if err := r.own(snapshotClass); err != nil {
+					return fmt.Errorf("failed to own VolumeSnapshotClass resource: %v", err)
+				}
 
-	if utils.AddAnnotation(r.storageClient, utils.DesiredConfigHashAnnotationKey, storageClientResponse.DesiredConfigHash) {
-		if err := r.update(r.storageClient); err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to update StorageClient with desired config hash annotation: %v", err)
+				if snapshotClass.Parameters == nil {
+					snapshotClass.Parameters = map[string]string{}
+				}
+				snapshotClass.Parameters["csi.storage.k8s.io/snapshotter-secret-namespace"] = r.OperatorNamespace
+
+				utils.AddLabels(snapshotClass, eResource.Labels)
+				utils.AddAnnotations(snapshotClass, eResource.Annotations)
+				return nil
+			}); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to create or update VolumeSnapshotClass: %s", err)
+			}
+		case "VolumeGroupSnapshotClass":
+			// check for CRD availability
+			if r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+				volumeGroupSnapshotClass := &groupsnapapi.VolumeGroupSnapshotClass{}
+				volumeGroupSnapshotClass.Name = eResource.Name
+				err = utils.CreateOrReplace(r.ctx, r.Client, volumeGroupSnapshotClass, func() error {
+					err = json.Unmarshal(eResource.Data, &volumeGroupSnapshotClass)
+					if err != nil {
+						return fmt.Errorf("failed to unmarshal VolumeGroupSnapshotClass configuration response: %v", err)
+					}
+					volumeGroupSnapshotClass.Name = eResource.Name
+					removeStorageClaimAsOwner(volumeGroupSnapshotClass)
+					if err := r.own(volumeGroupSnapshotClass); err != nil {
+						return fmt.Errorf("failed to own VolumeGroupSnapshotClass resource: %v", err)
+					}
+
+					if volumeGroupSnapshotClass.Parameters == nil {
+						volumeGroupSnapshotClass.Parameters = map[string]string{}
+					}
+					volumeGroupSnapshotClass.Parameters["csi.storage.k8s.io/group-snapshotter-secret-namespace"] = r.OperatorNamespace
```

Comment:
Why are we adding these parameters here and not on the provider?

Reply:
#323 is the client side fix and provider side will be done at a later point.